### DISCUSSION TITLE
Return in macro quotations

### DISCUSSION THREAD

**Discussion by cquiroz** [scala/scala3#13494] [Metaprogramming]
While trying to port `spire` macros to scala 3 I found that it doesn't seem possible to have a return clause in inline. If I try to do something like
```scala
'{
   return blah
}
```
I get an error saying `No explicit return allowed from inlineable`
As a bit of context I need this to allow functions calling the macro to return early, like
```scala
def foo = {
  A
  bar
  B
}

inline def bar = {
  C
  if (cond) return blah
  D
}
```
becoming
```scala
def foo = {
  A
  C
  return blah
}
```

**Comment by nicolasstucki** [comment]
That will be impossible with quasiquotes. It will require the use of the reflection API to create the `return`.

What you are trying to encode is not really a return, it is a non-local return (`longjmp` in C terms). We provide `scala.util.control.NonLocalReturns` for this purpose.

**Reply by cquiroz** [reply]
That's interesting, I never got into that side of the macros, I'll investigate
